File: ascending-data/InsertionSort/InsertionSort.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"time"
)

func insertionSort(arr []int) {
	for i := 1; i < len(arr); i++ {
		key := arr[i]
		j := i - 1
		for j >= 0 && arr[j] > key {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = key
	}
}

func main() {
	data := readData("ascending_data.txt")
	start := time.Now()
	insertionSort(data)
	elapsed := time.Since(start)
	fmt.Printf("Insertion Sort Execution Time: %.6f seconds\n", elapsed.Seconds())
}

func readData(filename string) []int {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var data []int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		num, _ := strconv.Atoi(scanner.Text())
		data = append(data, num)
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}
	return data
}

File: ascending-data/InsertionSort/InsertionSort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
void insertionSort(int arr[], int n) {
    int i, key, j;
    for (i = 1; i < n; i++) {
        key = arr[i];
        j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}
int main() {
    FILE *file = fopen("ascending_data.txt", "r");
    if (file == NULL) {
        printf("Error opening file!\n");
        return 1;
    }
    int arr[25000], n = 0;
    while (fscanf(file, "%d", &arr[n]) != EOF) {
        n++;
    }
    fclose(file);
    clock_t start, end;
    start = clock();
    insertionSort(arr, n);
    end = clock();
    double time_taken = ((double)(end - start)) / CLOCKS_PER_SEC;
    printf("Insertion Sort Execution Time: %.4f milliseconds\n", time_taken * 1000);
    return 0;
}